sw/otbn/crypto/div.s

```diff
@@ -0,0 +1,369 @@
+    jal       x0, _cond_sub_shifted_loop_end
+
+    _cond_sub_shifted_high_limb:
+    /* FG0.C <= 0 < (y[k] + FG0.C) */
+    bn.cmpb   w31, w24
+
+    _cond_sub_shifted_loop_end:
+    nop
+
+  /* Compare with the last limb of y (stored in a register).
+       FG0.C <= 0 < (y_upper + FG0.C) */
+  bn.cmpb   w31, w27
+
+  /* Compute a mask based on the final borrow.
+       w23 <= FG0.C ? 2^256 - 1 : 0 = mask */
+  bn.subb   w23, w31, w31
+
+  /* Get a pointer to the ith limb of r again.
+       x2 <= x10 + x3 = dptr_r + i*32 */
+  add      x2, x3, x10
+
+  /* Calculate the number of iterations for the copy loop. Since i < n, this
+     will always be nonzero.
+       x4 <= x9 - x8 = n - i */
+  sub       x5, x9, x8
+
+  /* Clear flags. */
+  bn.add    w31, w31, w31
+
+  /* Add back the low limbs of y if there was underflow. */
+  addi      x3, x11, 0
+  loop      x5, 5
+    /* w24 <= r[k+i] */
+    bn.lid    x24, 0(x2)
+    /* w25 <= y[k] */
+    bn.lid    x25, 0(x3++)
+    /* w25 <= y[k] & mask */
+    bn.and    w25, w25, w23
+    /* w24 <= r[k+i] + (y[k] & mask) */
+    bn.addc   w24, w24, w25
+    /* r[k+i] <= w24 */
+    bn.sid    x24, 0(x2++)
+
+  ret
+
+/**
+ * Constant-time floor division of arbitrarily large numbers.
+ *
+ * Returns q = floor(x / y).
+ *
+ * This routine performs binary long division. Faster algorithms certainly
+ * exist, but this one is comparatively simple to write and to make
+ * constant-time. In pseudocode:
+ *   q = 0 // quotient
+ *   r = x // remainder
+ *   for i=nbits(x)..0:
+ *     q <<= 1
+ *     if r >= (y << i)
+ *       r -= (y << i)
+ *       q += 1
+ *
+ * The general algorithmic approach is inspired by BoringSSL's implementation:
+ *   https://boringssl.googlesource.com/boringssl/+/1b2b7b2e70ce5ff50df917ee7745403d824155c5/crypto/fipsmodule/bn/div.c#457
+ *
+ * None of the caller-provided buffers may overlap in DMEM. The buffer for the
+ * numerator will be overwritten with the remainder, while the denominator will
+ * be preserved as-is.
+ *
+ * Flags have no meaning outside the scope of this subroutine.
+ *
+ * @param[in]  x9: n, number of 256-bit limbs for inputs x and y
+ * @param[in] x10: dptr_x, pointer to numerator x in dmem
+ * @param[in] x11: dptr_y, pointer to denominator y in dmem
+ * @param[in] x12: dptr_q, pointer to buffer for quotient q in dmem
+ * @param[in] w31: all-zero
+ * @param[out] dmem[dptr_q..dptr_q+n*32]: q, quotient
+ * @param[out] dmem[dptr_x..dptr_x+n*32]: r, remainder
+ *
+ * clobbered registers: x2 to x5, x8, x23 to x25, w23 to w27
+ * clobbered flag groups: FG0
+ */
+.globl div
+div:
+  /* Initialize quotient to zero.
+       dmem[dptr_q..dptr_q+n*32] = 0  */
+  addi      x2, x12, 0
+  li        x3, 31
+  loop      x9, 1
+    bn.sid    x3, 0(x2++)
+
+  /* Initialize loop counter.
+       x8 <= x9 = n */
+  addi      x8, x9, 0
+
+  /* Initialize constants for loop.
+       x23 = 23
+       x24 = 24
+       x25 = 25
+       x26 = 26 */
+  li        x23, 23
```

Comment:
All these constants made my life easier reviewing this code, specially the convention of mapping x[num] to w[num].